internal/api/impl/proxy/proxy.go

```diff
@@ -28,67 +29,187 @@ import (
 	"github.com/perses/perses/internal/api/interface/v1/globaldatasource"
 	"github.com/perses/perses/internal/api/interface/v1/globalsecret"
 	"github.com/perses/perses/internal/api/interface/v1/secret"
+	"github.com/perses/perses/internal/api/shared"
 	"github.com/perses/perses/internal/api/shared/crypto"
 	databaseModel "github.com/perses/perses/internal/api/shared/database/model"
+	"github.com/perses/perses/internal/api/shared/rbac"
 	"github.com/perses/perses/internal/api/shared/route"
 	"github.com/perses/perses/internal/api/shared/utils"
 	v1 "github.com/perses/perses/pkg/model/api/v1"
 	datasourceHTTP "github.com/perses/perses/pkg/model/api/v1/datasource/http"
+	"github.com/perses/perses/pkg/model/api/v1/role"
 	promConfig "github.com/prometheus/common/config"
 	"github.com/sirupsen/logrus"
 )
 
+// unsavedProxyBody is the body of the request when the datasource is not saved yet.
+// It contains the body of the request and the datasource spec, which is used to build the proxy rather than
+// retrieving the datasource from the database.
+type unsavedProxyBody struct {
+	Method string             `json:"method" yaml:"method"`
+	Body   []byte             `json:"body" yaml:"body"`
+	Spec   *v1.DatasourceSpec `json:"spec" yaml:"spec"`
+}
+
+func (u *unsavedProxyBody) setRequestParams(ctx echo.Context) {
+	req := ctx.Request()
+	req.Method = u.Method
+
+	if len(u.Body) > 0 {
+		req.Body = io.NopCloser(strings.NewReader(string(u.Body)))
+	} else {
+		req.Body = nil
+	}
+
+	ctx.SetRequest(req)
+}
+
 type endpoint struct {
 	Dashboard    dashboard.DAO
 	Secret       secret.DAO
 	GlobalSecret globalsecret.DAO
 	DTS          datasource.DAO
 	GlobalDTS    globaldatasource.DAO
 	Crypto       crypto.Crypto
+
+	rbac rbac.RBAC
 }
 
-func New(dashboardDAO dashboard.DAO, secretDAO secret.DAO, globalSecretDAO globalsecret.DAO, dtsDAO datasource.DAO, globalDtsDAO globaldatasource.DAO, crypto crypto.Crypto) route.Endpoint {
+func New(dashboardDAO dashboard.DAO, secretDAO secret.DAO, globalSecretDAO globalsecret.DAO, dtsDAO datasource.DAO, globalDtsDAO globaldatasource.DAO, crypto crypto.Crypto, rbac rbac.RBAC) route.Endpoint {
 	return &endpoint{
 		Dashboard:    dashboardDAO,
 		Secret:       secretDAO,
 		GlobalSecret: globalSecretDAO,
 		DTS:          dtsDAO,
 		GlobalDTS:    globalDtsDAO,
 		Crypto:       crypto,
+
+		rbac: rbac,
 	}
 }
 
+func (e *endpoint) checkPermission(ctx echo.Context, projectName string, scopeKind v1.Kind, action role.Action) error {
```

Comment:
instead of using the `v1.Kind` for `scopeKind`, why not using directly the `role.Scope` that match the type of datasource we want to use ? https://github.com/perses/perses/blob/main/pkg/model/api/v1/role/scope.go#L28

Like that we would avoid the usage of the method `role.GetScope`.

Reply:
I was sort of assuming that `role.GetScope` did something a bit more complicated 😅 Done